Source/WebCore/dom/InternalObserverReduce.h

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2024 Marais Rossouw <me@marais.co>. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Forward.h>
+
+namespace WebCore {
+
+class DeferredPromise;
+class Observable;
+class ReducerCallback;
+class ScriptExecutionContext;
+struct SubscribeOptions;
+
+void createInternalObserverOperatorReduce(ScriptExecutionContext&, Observable&, Ref<ReducerCallback>&&, const std::optional<JSC::JSValue>&, const SubscribeOptions&, Ref<DeferredPromise>&&);
```

Comment:
I think this can take `JSC::JSValue` instead of `const std::optional<JSC::JSValue>&`. Not sure that the optional is really working for us here. Could just use `undefined` at first maybe? I guess we can leave this as-is for now while we create test cases that try to tickle `undefined` vs. no value for initial value.

Reply:
I think this best explains it: https://github.com/WebKit/WebKit/pull/37630#discussion_r1876908305

We dont want to call the reducer if the initial value is undefined/not-passed. There is a certain convince imposed by std::optional.

But I also hear you, perhaps its better we localize this logic closer to where it matters. Ill move it.

Reply:
Okay pushed a change that moves this down to where it matters. Thanks for the nudge!